File: android/core/protocol/src/main/kotlin/app/hapi/protocol/chat/Normalize.kt
package app.hapi.protocol.chat

import app.hapi.protocol.wire.DecryptedMessage

/** Port of `web/src/chat/normalize.ts` — the decode-tree entry point. */

private fun stringifiedAgentFallback(
    message: DecryptedMessage,
    text: String,
    meta: kotlinx.serialization.json.JsonElement?,
    invokedAt: Long?,
): NormalizedMessage = NormalizedMessage.Agent(
    id = message.id,
    localId = message.localId,
    createdAt = message.createdAt,
    isSidechain = false,
    content = listOf(NormalizedAgentContent.Text(text = text, uuid = message.id, parentUUID = null)),
    meta = meta,
    status = null,
    originalText = null,
    invokedAt = invokedAt,
)

fun normalizeDecryptedMessage(message: DecryptedMessage): NormalizedMessage? {
    val record = unwrapRoleWrappedRecordEnvelope(message.content)
    if (record == null) {
        // No envelope: stringify the whole content. (TS sets no invokedAt here.)
        return stringifiedAgentFallback(message, safeStringify(message.content), meta = null, invokedAt = null)
    }

    val invokedAt = message.invokedAtOrNull

    if (record.role == "user") {
        val normalized = normalizeUserRecord(message.id, message.localId, message.createdAt, record.content, record.meta)
        return if (normalized is NormalizedMessage.User) {
            normalized.copy(invokedAt = invokedAt)
        } else {
            NormalizedMessage.User(
                id = message.id,
                localId = message.localId,
                createdAt = message.createdAt,
                isSidechain = false,
                text = safeStringify(record.content),
                meta = record.meta,
                invokedAt = invokedAt,
            )
        }
    }
    if (record.role == "agent") {
        if (isSkippableAgentContent(record.content)) {
            return null
        }
        val normalized = normalizeAgentRecord(message.id, message.localId, message.createdAt, record.content, record.meta)
        if (normalized == null && isCodexContent(record.content)) {
            return null
        }
        return when (normalized) {
            is NormalizedMessage.User -> normalized.copy(invokedAt = invokedAt)
            is NormalizedMessage.Agent -> normalized.copy(invokedAt = invokedAt)
            is NormalizedMessage.Event -> normalized.copy(invokedAt = invokedAt)
            null -> stringifiedAgentFallback(message, safeStringify(record.content), meta = record.meta, invokedAt = invokedAt)
        }
    }

    return stringifiedAgentFallback(message, safeStringify(record.content), meta = record.meta, invokedAt = invokedAt)
}

File: ios/Packages/HapiKit/Sources/HapiProtocol/Chat/Normalize.swift
import Foundation

// Port of web/src/chat/normalize.ts — total decode of one wire message.

/// Port of `normalizeDecryptedMessage`. Returns `nil` only on the two
/// legitimate drop paths (skippable Claude output, unknown codex content);
/// every other malformed shape degrades to stringified text.
public func normalizeDecryptedMessage(_ message: DecryptedMessage) -> NormalizedMessage? {
    guard let record = unwrapRoleWrappedRecordEnvelope(message.content) else {
        return NormalizedMessage(
            id: message.id,
            localId: message.localId,
            createdAt: message.createdAt,
            content: .agent([.text(.init(text: safeStringify(message.content), uuid: message.id, parentUUID: nil))]),
            isSidechain: false
            // No meta / invokedAt on this branch (normalize.ts:10-20).
        )
    }

    if record.role == "user" {
        if var normalized = normalizeUserRecord(
            messageId: message.id,
            localId: message.localId,
            createdAt: message.createdAt,
            content: record.content,
            meta: record.meta
        ) {
            normalized.invokedAt = message.invokedAt
            return normalized
        }
        return NormalizedMessage(
            id: message.id,
            localId: message.localId,
            createdAt: message.createdAt,
            content: .user(text: safeStringify(record.content), attachments: nil),
            isSidechain: false,
            meta: record.meta,
            invokedAt: message.invokedAt
        )
    }

    if record.role == "agent" {
        if isSkippableAgentContent(record.content) {
            return nil
        }
        let normalized = normalizeAgentRecord(
            messageId: message.id,
            localId: message.localId,
            createdAt: message.createdAt,
            content: record.content,
            meta: record.meta
        )
        if normalized == nil && isCodexContent(record.content) {
            return nil
        }
        if var normalized {
            normalized.invokedAt = message.invokedAt
            return normalized
        }
        return NormalizedMessage(
            id: message.id,
            localId: message.localId,
            createdAt: message.createdAt,
            content: .agent([.text(.init(text: safeStringify(record.content), uuid: message.id, parentUUID: nil))]),
            isSidechain: false,
            meta: record.meta,
            invokedAt: message.invokedAt
        )
    }

    return NormalizedMessage(
        id: message.id,
        localId: message.localId,
        createdAt: message.createdAt,
        content: .agent([.text(.init(text: safeStringify(record.content), uuid: message.id, parentUUID: nil))]),
        isSidechain: false,
        meta: record.meta,
        invokedAt: message.invokedAt
    )
}
